tests/robustness/watch.go

```diff
@@ -45,37 +46,79 @@ func collectClusterWatchEvents(ctx context.Context, t *testing.T, lg *zap.Logger
 		if err != nil {
 			t.Fatal(err)
 		}
-
+		memberChan := make(chan int64, 1)
+		memberMaxRevisionChans[i] = memberChan
 		wg.Add(1)
 		go func(i int, c *clientv3.Client) {
 			defer wg.Done()
 			defer c.Close()
-			responses := watchMember(ctx, lg, c)
+			responses := watchMember(ctx, t, c, memberChan)
 			mux.Lock()
 			memberResponses[i] = responses
 			mux.Unlock()
 		}(i, c)
 	}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		maxRevision := <-maxRevisionChan
+		for _, memberChan := range memberMaxRevisionChans {
+			memberChan <- maxRevision
+		}
+	}()
 	wg.Wait()
 	return memberResponses
 }
 
-func watchMember(ctx context.Context, lg *zap.Logger, c *clientv3.Client) (resps []watchResponse) {
+// watchMember collects all responses until context is cancelled or has observed revision provided via maxRevisionChan.
+func watchMember(ctx context.Context, t *testing.T, c *clientv3.Client, maxRevisionChan <-chan int64) (resps []watchResponse) {
+	var maxRevision int64 = 0
 	var lastRevision int64 = 0
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	watch := c.Watch(ctx, "", clientv3.WithPrefix(), clientv3.WithRev(1), clientv3.WithProgressNotify())
 	for {
 		select {
 		case <-ctx.Done():
+			revision := watchResponsesMaxRevision(resps)
+			if maxRevision == 0 {
+				t.Errorf("Client didn't collect all events, max revision not set")
+			}
+			if revision < maxRevision {
+				t.Errorf("Client didn't collect all events, revision got %d, expected: %d", revision, maxRevision)
+			}
 			return resps
-		default:
+		case maxRevision = <-maxRevisionChan:
+			if lastRevision >= maxRevision {
+				cancel()
+			}
+		case resp := <-watch:
+			if resp.Err() == nil {
+				resps = append(resps, watchResponse{resp, time.Now()})
+			} else if !resp.Canceled {
+				t.Errorf("Watch stream received error, err %v", resp.Err())
+			}
```

Comment:
Theoretically, after `cancel()`, will `resp := <-watch` wins against `case <-ctx.Done()`  and error out due to `resp.Canceled = true`?

Reply:
On `cancel()` I expect that watch will get a response with `closeErr` equal `cancelled` and `Cancelled` equal `true`.
 So I expect this if statement to not error as we only error on `!resp.Canceled` so cancel will never cause error.

Reply:
Thanks! I must misread the `if !resp.Canceled` condition check in the opposite way.